### PR TITLE
Update AssemblyBuilder documentation

### LINKED ISSUE #5169: System.Reflection.Emit.AssemblyBuilder no longer supports Save on NET Core
Document file: [AssemblyBuilder.xml](https://github.com/dotnet/dotnet-api-docs/blob/7e80efdc53be82550e4a304d95cf0057710a3936/xml/System.Reflection.Emit/AssemblyBuilder.xml)

The example code is using a removed feature.

### LINKED ISSUE #6717: AssemblyBuilder examples don't compile on .NET 5.0 (and probably any version of Core)
**Issue description**

https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.assemblybuilder?view=net-5.0#examples does not compile in .NET 5.0 as the APIs for emitting code to output files are not yet complete. The example should be removed for non-Framework versions of that page.

### LINKED ISSUE #6889: Update docs for AssemblyBuilder for net core, net standard, net 5 and higher
---
**Issue moved from MicrosoftDocs/feedback#3655**
- Please respond to @snaulX.

---

_From @snaulX on Sunday, July 4, 2021 2:22:00 PM_

All of this docs of [AssemblerBuilder](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.assemblybuilder) for .NET Core and higher contains info that you can **save** assembly using `AssemblyBuilderAccess.RunAndSave` or `AssemblyBuilderAccess.Save`. But these access modes supports only in .NET Framework. So fix sample and remove or edit information about saving assemblies on platforms that higher than .NET Framework.

### PR DESCRIPTION
**Pull request by MSDN-WhiteKnight**
- Add information that saving to files is only supported in .NET Framework
- Add information that multiple modules, remoting, and app domains are only relevant in .NET Framework
- Replace AppDomain.DefineDynamicAssembly with AssemblyBuilder.DefineDynamicAssembly available in new runtimes
- Update examples to avoid APIs not available in new runtimes

Fixes #6889  
Fixes #6717  
Fixes #5169  
Contributes to #5321, #2645, #4207

### CODE REVIEW

**Comment by BillWagner** [comment]
ping @carlossanlop to route this for review

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 1]
```diff
@@ -68,55 +68,47 @@
...
+ 
+In .NET Framework, it is possible to save dynamic assemblies to files. This feature is not available in .NET Core and .NET 5. See <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder> for an alternative way to generate assembly files.
+  
+In .NET Framework, a dynamic assembly can consist of one or more dynamic modules. If a dynamic assembly contains more than one dynamic module, the assembly's manifest file name should match the module's name that is specified as the first argument to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicModule%2A> method. In .NET Core and .NET 5, a dynamic assembly can only consist of one dynamic module. 
+
+### Persistable dynamic assemblies in .NET Framework
+In .NET Framework, dynamic assemblies and modules could be saved to files. To support this feature, <xref:System.Reflection.Emit.AssemblyBuilderAccess> enumeration declares additional values, <xref:System.Reflection.Emit.AssemblyBuilderAccess.Run> and <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndSave>. The dynamic assembly created using one of these access modes is called a persistable assembly, while the regular memory-only assembly is called transient.
```
Comment (steveharter) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:93]:
```suggestion
In .NET Framework, dynamic assemblies and modules can be saved to files. To support this feature, <xref:System.Reflection.Emit.AssemblyBuilderAccess> enumeration declares additional values, <xref:System.Reflection.Emit.AssemblyBuilderAccess.Run> and <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndSave>. The dynamic assembly created using one of these access modes is called a persistable assembly, while the regular memory-only assembly is called transient.
```

**Review by steveharter — COMMENTED** [review]
Thanks for the updates.

I haven't been following .NET Framework vs. .NET (core) doc changes. It appears there is only one set of doc for both, with .NET Framework having basically footnotes for the extra functionality?

**Comment by gewarren** [comment]
> Thanks for the updates.
> 
> I haven't been following .NET Framework vs. .NET (core) doc changes. It appears there is only one set of doc for both, with .NET Framework having basically footnotes for the extra functionality?

Yes, in the API docs, the doc comments are shared between .NET Framework and .NET (Core).

**Review by tdykstra — COMMENTED** [review]
LGTM

**Comment by MSDN-WhiteKnight** [comment]
>It appears there is only one set of doc for both, with .NET Framework having basically footnotes for the extra functionality?

Yes, that's the idea. Because .NET Core reflection emit functionality is a subset of .NET Framework's, it's best to have the main part, describing functionality shared by both implementations, and additional .NET Framework-specific notes.

**Comment by MSDN-WhiteKnight** [comment]
The CI build complains about the absence of project files for snippets. I've added projects for C# and VB, but the C++/CLI is a bit more complicated, because it is not supported by .NET SDK, and requires VS with C++/CLI component installed. Are project files for C++/CLI snippet required for merge here?

**Comment by gewarren** [comment]
> The CI build complains about the absence of project files for snippets. I've added projects for C# and VB, but the C++/CLI is a bit more complicated, because it is not supported by .NET SDK, and requires VS with C++/CLI component installed. Are project files for C++/CLI snippet required for merge here?

@adegeo Could you answer this one?

**Comment by adegeo** [comment]
@MSDN-WhiteKnight When you want the snippets system to run VS, add a `snippets.5000.json` file to the same folder as the project with the following content:

```json
{
    "host": "visualstudio"
}
```

**Comment by MSDN-WhiteKnight** [comment]
Added C++/CLI project, CI build passes now.

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 2]
```diff
@@ -68,55 +68,47 @@
...
+     The dynamic assembly represented by an <xref:System.Reflection.Emit.AssemblyBuilder> can be used to execute the emitted code.
+  
+-   <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndCollect?displayProperty=nameWithType>  
+  
+     The dynamic assembly represented by an <xref:System.Reflection.Emit.AssemblyBuilder> can be used to execute the emitted code and is automatically reclaimed by garbage collector.  
+
+  
+The access mode must be specified by providing the appropriate <xref:System.Reflection.Emit.AssemblyBuilderAccess> value in the call to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicAssembly%2A?displayProperty=nameWithType> method when the dynamic assembly is defined and cannot be changed later. The runtime uses the access mode of a dynamic assembly to optimize the assembly's internal representation.
+ 
+In .NET Framework, it is possible to save dynamic assemblies to files. This feature is not available in .NET Core and .NET 5. See <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder> for an alternative way to generate assembly files.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:88]:
```suggestion
In .NET Framework, you can save dynamic assemblies to files. This feature is not available in .NET Core and .NET 5 and later versions.  For an alternative way to generate assembly files, see <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder>.
```

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 3]
```diff
@@ -68,55 +68,47 @@
...
+  
+In .NET Framework, a dynamic assembly can consist of one or more dynamic modules. If a dynamic assembly contains more than one dynamic module, the assembly's manifest file name should match the module's name that is specified as the first argument to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicModule%2A> method. In .NET Core and .NET 5, a dynamic assembly can only consist of one dynamic module.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:90]:
```suggestion
In .NET Framework, a dynamic assembly can consist of one or more dynamic modules. If a dynamic assembly contains more than one dynamic module, the assembly's manifest file name should match the module's name that is specified as the first argument to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicModule%2A> method. In .NET Core and .NET 5+, a dynamic assembly can only consist of one dynamic module. 
```

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 4]
```diff
@@ -68,55 +68,47 @@
...
+ 
+In .NET Framework, it is possible to save dynamic assemblies to files. This feature is not available in .NET Core and .NET 5. See <xref:System.Reflection.Metadata.Ecma335.MetadataBuilder> for an alternative way to generate assembly files.
+  
+In .NET Framework, a dynamic assembly can consist of one or more dynamic modules. If a dynamic assembly contains more than one dynamic module, the assembly's manifest file name should match the module's name that is specified as the first argument to the <xref:System.Reflection.Emit.AssemblyBuilder.DefineDynamicModule%2A> method. In .NET Core and .NET 5, a dynamic assembly can only consist of one dynamic module. 
+
+### Persistable dynamic assemblies in .NET Framework
+In .NET Framework, dynamic assemblies and modules can be saved to files. To support this feature, <xref:System.Reflection.Emit.AssemblyBuilderAccess> enumeration declares additional values, <xref:System.Reflection.Emit.AssemblyBuilderAccess.Run> and <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndSave>. The dynamic assembly created using one of these access modes is called a persistable assembly, while the regular memory-only assembly is called transient.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:93]:
```suggestion

In .NET Framework, dynamic assemblies and modules can be saved to files. To support this feature, the <xref:System.Reflection.Emit.AssemblyBuilderAccess> enumeration declares two additional fields: <xref:System.Reflection.Emit.AssemblyBuilderAccess.Run> and <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndSave>. The dynamic assembly created using one of these access modes is called a *persistable* assembly, while the regular memory-only assembly is called *transient*.
```

File: xml/System.Reflection.Emit/AssemblyBuilder.xml [hunk 5]
```diff
@@ -68,55 +68,47 @@
...
   
 ## Examples  
- The following code example shows how to define a dynamic assembly with one module. The module in the example assembly contains one type, `MyDynamicType`, which has a private field, a property that gets and sets the private field, constructors that initialize the private field, and a method that multiplies a user-supplied number by the private field value and returns the result.  
-  
- <xref:System.Reflection.Emit.AssemblyBuilderAccess.RunAndSave?displayProperty=nameWithType> is specified when the assembly is created. The assembly code is used immediately, and the assembly is also saved to disk so that it can be examined with [Ildasm.exe (IL Disassembler)](/dotnet/framework/tools/ildasm-exe-il-disassembler) or used in another program.  
-  
+The following code example shows how to define and use a dynamic assembly. The example assembly contains one type, `MyDynamicType`, which has a private field, a property that gets and sets the private field, constructors that initialize the private field, and a method that multiplies a user-supplied number by the private field value and returns the result.
```
Comment (gewarren) [inline xml/System.Reflection.Emit/AssemblyBuilder.xml:107]:
```suggestion
The following code example shows how to define and use a dynamic assembly. The example assembly contains one type, `MyDynamicType`, that has a private field, a property that gets and sets the private field, constructors that initialize the private field, and a method that multiplies a user-supplied number by the private field value and returns the result.
```

**Review by gewarren — APPROVED** [review]

**Review by steveharter — APPROVED** [review]

**Pull request merged**